### PR TITLE
Update implementation of DNSSEC03

### PR DESCRIPTION
**Pull request by tgreenx**
## Purpose

This PR proposes an update to the implementation of DNSSEC03 following an update to the specification (https://github.com/zonemaster/zonemaster/pull/1189).

Note that Prefix Suffix List check (step 13.2.2.1) is not yet implemented, as this service is not yet provided by Zonemaster.

Requires a fix to Zonemaster::LDNS::RR::NSEC3 to work: https://github.com/zonemaster/zonemaster-ldns/pull/177

## Context

Test Case specification: https://github.com/zonemaster/zonemaster/pull/1189
Test Zones specification: https://github.com/zonemaster/zonemaster/pull/1218

Requires https://github.com/zonemaster/zonemaster-ldns/pull/177

## Changes

`lib/Zonemaster/Engine/Test/DNSSEC.pm`

- Update implementation
- Update message tags and message ids

`share/profile.json` and `share/profile.yaml`

- Update message tags

`t/Test-dnssec03.t` and `t/Test-dnssec03.data`

- Update unitary tests implementation and data

## How to test this PR

Unit tests are updated and should pass.

**Manual testing:**
```
$ git log -1 --oneline
25036e6 (HEAD -> update-dnssec03, origin/update-dnssec03) Update implementation of DNSSEC03

$ zonemaster-cli --test dnssec/dnssec03 --show-testcase --level INFO zonemaster.net
Seconds Level    Testcase       Message
======= ======== ============== =======
   0.01 INFO     UNSPECIFIED    Using version v4.7.3 of the Zonemaster engine.
   5.79 INFO     DNSSEC03       The zone does not use NSEC3. Testing for NSEC3 has been skipped. Fetched from name servers "ns2.nic.fr/192.93.0.4;ns2.nic.fr/2001:660:3005:1::1:2;nsa.dnsnode.net/194.58.192.46;nsa.dnsnode.net/2a01:3f1:46::53;nsp.dnsnode.net/194.58.198.32;nsp.dnsnode.net/2a01:3f1:3032::53;nsu.dnsnode.net/185.42.137.98;nsu.dnsnode.net/2a01:3f0:400::32".

$ zonemaster-cli --test dnssec/dnssec03 --show-testcase --level INFO afnic.fr
Seconds Level    Testcase       Message
======= ======== ============== =======
   0.00 INFO     UNSPECIFIED    Using version v4.7.3 of the Zonemaster engine.
   3.66 INFO     DNSSEC03       The following servers respond with a legal hash algorithm in NSEC3. Fetched from name servers "g.ext.nic.fr/194.0.36.1;g.ext.nic.fr/2001:678:4c::1;ns1.nic.fr/192.134.4.1;ns1.nic.fr/2001:67c:2218:2::4:1;ns2.nic.fr/192.93.0.4;ns2.nic.fr/2001:660:3005:1::1:2;ns3.nic.fr/192.134.0.49;ns3.nic.fr/2001:660:3006:1::1:1".
   3.66 NOTICE   DNSSEC03       The following servers respond with NSEC3 opt-out enabled. The recommended practice is to disable opt-out. Fetched from name servers "g.ext.nic.fr/194.0.36.1;g.ext.nic.fr/2001:678:4c::1;ns1.nic.fr/192.134.4.1;ns1.nic.fr/2001:67c:2218:2::4:1;ns2.nic.fr/192.93.0.4;ns2.nic.fr/2001:660:3005:1::1:2;ns3.nic.fr/192.134.0.49;ns3.nic.fr/2001:660:3006:1::1:1".
   3.66 INFO     DNSSEC03       The following servers respond with NSEC3 iteration value set to zero (as recommended). Fetched from name servers "g.ext.nic.fr/194.0.36.1;g.ext.nic.fr/2001:678:4c::1;ns1.nic.fr/192.134.4.1;ns1.nic.fr/2001:67c:2218:2::4:1;ns2.nic.fr/192.93.0.4;ns2.nic.fr/2001:660:3005:1::1:2;ns3.nic.fr/192.134.0.49;ns3.nic.fr/2001:660:3006:1::1:1".
   3.66 INFO     DNSSEC03       The following servers respond with a legal empty salt in NSEC3. Fetched from name servers "g.ext.nic.fr/194.0.36.1;g.ext.nic.fr/2001:678:4c::1;ns1.nic.fr/192.134.4.1;ns1.nic.fr/2001:67c:2218:2::4:1;ns2.nic.fr/192.93.0.4;ns2.nic.fr/2001:660:3005:1::1:2;ns3.nic.fr/192.134.0.49;ns3.nic.fr/2001:660:3006:1::1:1".
```

### CODE REVIEW

**Comment by matsduf** [comment]
"Not in MANIFEST: t/Test-dnssec03.t"

**Comment by matsduf** [comment]
> "/usr/local/bin/perl" "-Iinc" "-MExtUtils::Manifest=fullcheck" -e fullcheck
> No such file: t/Test-dnssec03.data

**Comment by matsduf** [comment]
```
$ zonemaster-cli --raw  --test DNSSEC/dnssec03 --hints COMMON/hintfile --level info bad-values.dnssec03.xa
   0.00 INFO     GLOBAL_VERSION   version=v4.7.3
free(): double free detected in tcache 2
Aborted (core dumped)
```

**Comment by matsduf** [comment]
```
$ zonemaster-cli --raw  --test DNSSEC/dnssec03 --hints COMMON/hintfile --level info inconsistent-values.dnssec03.xa
   0.00 INFO     GLOBAL_VERSION   version=v4.7.3
free(): double free detected in tcache 2
Aborted (core dumped)
```

**Comment by matsduf** [comment]
In two cases `zonemaster-cli` crashes (see above). In the test zones manipulated NSEC3 records are used, and it might be that LDNS does not like that. Maybe I have to create "real" NSEC3 records for the tests, but in any case Zonemaster should not crash.

**Comment by matsduf** [comment]
Must be updated:

```
    DNSSEC03 => sub {
        __x    # DNSSEC:DNSSEC03
          "Check for too many NSEC3 iterations";
    },
```
(https://github.com/zonemaster/zonemaster-engine/blob/be1bde54d0dae23e4a534dcf5966528697b93d7e/lib/Zonemaster/Engine/Test/DNSSEC.pm#L565C1-L568C6)

**Comment by matsduf** [comment]
This PR will then require updatd Zonemaster-LDNS? Will that affect Engine elsewhere?

**Comment by matsduf** [comment]
From scenario UNASSIGNED-FLAG-USED:
```
   0.18 ERROR    The following servers respond with an NSEC3 record where an unassigned flag is used (flag 2). Fetched from name servers "(...)".
```

The value of the flag integer is 2, but the flag, in this case, is 6, isn't it?

**Review by matsduf — APPROVED** [review]
Looks fine except maybe a detail from scenario UNASSIGNED-FLAG-USED (see my comment).

I have not checked why the two tests in `t/Test-dnssec.t` fails. Can they just be removed?

File: lib/Zonemaster/Engine/Test/DNSSEC.pm [hunk 1]
```diff
@@ -782,6 +794,115 @@ Readonly my %TAG_DESCRIPTIONS => (
...
+    DS03_SERVER_NO_NSEC3 => sub {
+        __x    # DNSSEC:DS03_SERVER_NO_NSEC3
+          'The following name servers do not use NSEC3, but others do. '
+          . 'Testing for NSEC3 has been skipped on the following servers. '
+          . 'Fetched from name servers "{ns_list}".',
+          @_;
+    },
+    DS03_UNASSIGNED_FLAG_USED => sub {
+        __x    # DNSSEC:DS03_UNASSIGNED_FLAG_USED
+          'The following servers respond with an NSEC3 record where an unassigned flag is used (flag {int}). '
```
Comment (marc-vanderwal) [inline lib/Zonemaster/Engine/Test/DNSSEC.pm:902]: May I suggest a different wording for this error message?
```suggestion
          'The following servers respond with an NSEC3 record where an unassigned flag is set at bit position {int}. '
```

**Comment by matsduf** [comment]
@tgreenx, please approve https://github.com/zonemaster/zonemaster-ldns/pull/177 and ask @marc-vanderwal to merge. I have approved this PR so for me it is fine, but I do not think it should be merged before we can see that CI passes.

**Comment by marc-vanderwal** [comment]
Alright, zonemaster/zonemaster-ldns#177 is merged, and I’m glad to see that all is well.

File: lib/Zonemaster/Engine/Test/DNSSEC.pm [hunk 2]
```diff
@@ -782,6 +794,115 @@ Readonly my %TAG_DESCRIPTIONS => (
...
+          . 'Fetched from name servers "{ns_list}".',
```
Comment (marc-vanderwal) [inline lib/Zonemaster/Engine/Test/DNSSEC.pm:903]: I think the wording of that error message has room for improvement; but I’ll comment on zonemaster/zonemaster#1189 instead.

**Comment by matsduf** [comment]
@tgreenx, https://github.com/zonemaster/zonemaster/pull/1189 has been updated and all review comments should be taken care of.

The test scenarios and test zones have been updated in https://github.com/zonemaster/zonemaster/pull/1218 to match the added test tags. Please note that all scenarios have been updated when it comes to "forbidden" tags, i.e. the new tags have been added to all old scenarios as "forbidden". Three new scenarios have been created and also test zones for them. As far as I can see the new test zones are correct.

**Review by matsduf — APPROVED** [review]

**Review by marc-vanderwal — APPROVED** [review]

**Pull request merged**